builder/dockerfile/command/commands.go

```diff
@@ -0,0 +1,232 @@
+package command
+
+import (
+	"errors"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/strslice"
+)
+
+type KeyValuePair struct {
+	Key   string
+	Value string
+}
+
+type KeyValuePairs []KeyValuePair
+
+// WithSourceCode is a marker indicating that a given command
+// Has been parsed from source code (wich can be displayed in builder output)
+type WithSourceCode interface {
+	SourceCode() string
+}
+
+type EnvCommand struct {
+	OriginalSource string
+	Env            KeyValuePairs // kvp slice instead of map to preserve ordering
+}
+
+func (c *EnvCommand) SourceCode() string {
```

Comment:
Embedded struct can probably also return normalized command name that can be used by dispatchers.